torch/utils/bottleneck/__main__.py

```diff
@@ -0,0 +1,231 @@
+import argparse
+import cProfile
+import pstats
+import subprocess
+import sys
+import os
+import re
+
+try:
+    from StringIO import StringIO
+except ImportError:
+    from io import StringIO
+
+import torch
+from torch.autograd import profiler
+
+PY3 = sys.version_info >= (3, 0)
+
+
+def run(command):
+    """Returns (return-code, stdout, stderr)"""
+    p = subprocess.Popen(command, stdout=subprocess.PIPE,
+                         stderr=subprocess.PIPE, shell=True)
+    output, err = p.communicate()
+    rc = p.returncode
+    if PY3:
+        output = output.decode("ascii")
+        err = err.decode("ascii")
+    return (rc, output, err)
+
+
+def check_running_cuda_version():
+    (rc, out, err) = run('nvcc --version')
+    if rc is not 0:
+        return None
+    m = re.search(r'V(.*)$', out)
+    assert m is not None
+    return m.group(1)
+
+
+def check_pip_packages():
+    # People generally have `pip` as `pip` or `pip3`
+    def run_with_pip(pip):
+        (rc, out, err) = run(pip + ' list --format=legacy | grep torch')
+        if rc is 0:
+            return '`{}` list truncated output:\n{}'.format(pip, out)
+        return None
+
+    result = []
+    out = run_with_pip('pip')
+    if out is not None:
+        result.append(out)
+    out_pip3 = run_with_pip('pip3')
+    if out_pip3 is not None:
+        result.append(out_pip3)
+
+    return '\n'.join(result)
+
+
+def compiled_with_cuda():
+    if torch.version.cuda:
+        return 'compiled w/ CUDA {}'.format(torch.version.cuda)
+    return 'not compiled w/ CUDA'
+
+
+def run_env_analysis():
+    print('Running environment analysis...')
+    result = []
+
+    debug_str = ''
+    if torch.version.debug:
+        debug_str = ' DEBUG'
+    result.append('PyTorch {}{} {}'.format(
+        torch.__version__, debug_str,
+        compiled_with_cuda()))
+
+    avail = 'Running with python {}.{}, '.format(sys.version_info[0], sys.version_info[1])
+    if torch.cuda.is_available():
+        cuda = check_running_cuda_version()
+        if cuda is None:
+            cuda = ''
+        avail += 'CUDA {}'.format(cuda)
+    else:
+        avail += 'CUDA unavailable'
+    result.append(avail)
+
+    result.append('')
+
+    pip = check_pip_packages()
+    if pip is not None:
+        result.append(check_pip_packages())
+
+    return '\n'.join(result)
+
+
+def set_env_variables(gpu_device):
+    # Override CUDA_LAUNCH_BLOCKING by default for more accurate profiling.
+    os.environ['CUDA_LAUNCH_BLOCKING'] = '1'
```

Comment:
We really should use the CUDA mode of the profiler instead of CUDA_LAUNCH_BLOCKING. This will give a very skewed picture. It would be nice if we could run the script twice: first without CUDA-mode profiling enabled, then with CUDA-mode profiling enabled. If the execution times are similar for both (which means that CUDA-mode didn't add a lot of overhead), present only the CUDA-level result. Otherwise present both.

Reply:
That makes sense, I'll make the change